docs/src/nondiff_points.md

```diff
@@ -0,0 +1,155 @@
+For our purpose we basically never care which we are working with and so write sub/super-derivative.
+
+For a function $f$ at some point $x_0$ a sub/super-derivative is a real number $c$ such that there exists a open ball $\mathcal{B} \subset \mathrm{dom}(f)$ containing $x_0$,
+and for all points $z \in \mathcal{B}$ the following holds:
+
+$$\mathrm{sub -derivative:}\qquad f(z) - f(x_0) \ge c\,(z-x_0)$$
+
+$$\mathrm{super-derivative:}\qquad f(z) - f(x_0) \le c\,(z-x_0)$$
+
+We call the the set of all values $c$ the sub/super-differential at $x_0$.
+
+More informally: consider a plot of the function.
+The sub/super-differential at a point is the set of slopes of all lines you could draw touching that point, but with the lines either entirely above, or entirely below the curve.
+
+It is best illustrated with a figure:
+
+![plot showing subderiviatives](https://upload.wikimedia.org/wikipedia/commons/4/4e/Subderivative_illustration.png)
+
+In this figure a plot of a function is shown in blue.
+Two subtangent lines are plotted in red.
+Their slopes are sub/super-derivatives at $x_0$, and they are two elements of the subdifferential.
+If you flip it upside down, it would apply for the super-differential, with the lines being above the curve.
+
+### Some key things to note.
+
+For a function where the derivative is defined on both sides of the point:
+ - the derivative on each side is a sub/super-derivative at the point
+ - as is the mean of the derivative of each side
+ - in-fact the mean of any subset (including the whole sub/super-differential) of sub/super-derivatives is also a sub/super-derivative.
+ - if the derivative one one side is negative and the other positive then zero is a sub/super-derivative.
+
+For a function that is only defined on one side of the point, the sub/super-differential is the full set of real numbers.
+This by the subgradient convention leaves you free to chose *any* useful value.
+
+!!! info "Does AD always return a sub/super-derivative? No"
+    On consideration of this, one might be tempted to think that AD always returns a sub/super-derivative.
+    And that in normal cases it return the only sub/super-derivative i.e. the actual derivative; and in other case it picks one of the branches.
+    Thus all weirdness of AD disagreeing with calculus could be explained away in this way.
+    **This is not the case.**
+    As it is not necessarily true that all branches derivatives are correct sub/super-differentials for the function.
+    Consider the function from [Abadi and Plotkin, 2019](https://dl.acm.org/doi/10.1145/3371106):
+    `f(x::Float64) = x == 0.0 ? 0.0 : x`.
+    The correct derivative of this function is `1` at all points, but most ADs will say that at `x=0` the derivative is zero.
+    The fix for this, is to define a rule which *does* in fact return a sub/super-derivative.
+    So one could say a correctly functioning AD with all needed rules does always return a sub/super-differential.
```

Comment:
@wesselb  raised another case, that is far simpler .
Sometimes the derivative exists but the subdifferential doesn't.
For example at a saddle point, like `x->x^3` for `x=0`.|
are saddle points the only continous examples?

Is it correct to say either the derivative exists or the subderivative exists?

Reply:
How about `cbrt` at zero? This isn't locally convex in either direction. 

It seems the term "subderivative" is more closely tied to convex functions than I realised.

Reply:
I think cbrt is fine.
is domain is `x>=0`
Which means  the open ball we draw will also only have points: `z>=0`
and There are many choices of `c` such that `c*(z-0) >= cbrt(z-0)` or `c*(z-0) <= cbrt(z-0)`
Like I think any real number ?
Which is not particular helpful for deciding what to return, but it is freeing.

Reply:
No, `cbrt(x)` chooses the real branch at negative `x`, unlike `x^(1/3)`.

Reply:
TIL

Reply:
`cbrt(x)`
![image](https://user-images.githubusercontent.com/5127634/127499121-a8691a18-bc2b-4fc6-b0c2-d1ab1aa36abc.png)
derivative:
![image](https://user-images.githubusercontent.com/5127634/127499286-a9b2005d-283f-4382-b5cf-4d2a70161307.png)